lib/ret/media_resolver.ex

```diff
@@ -0,0 +1,225 @@
+    imgur_api_url = "https://imgur-apiv3.p.mashape.com/3/album/#{album_id}"
+    uri = image_uri_for_imgur_collection_api_url(imgur_api_url) || uri
+
+    {:commit, uri |> URI.to_string()}
+  end
+
+  defp resolve_non_video(%URI{} = uri, _root_host) do
+    # Fall back on og: tags
+    resp = retry_get_until_success(uri |> URI.to_string())
+
+    uri =
+      case resp.body |> OpenGraph.parse() do
+        %{video: video} -> video |> URI.parse()
+        %{image: image} -> image |> URI.parse()
+        _ -> uri
+      end
+
+    {:commit, uri |> URI.to_string()}
+  end
+
+  defp image_uri_for_imgur_collection_api_url(imgur_api_url) do
+    with headers when is_list(headers) <- get_imgur_headers() do
+      imgur_resp = retry_get_until_success(imgur_api_url, headers)
+
+      image_id =
+        imgur_resp.body
+        |> Poison.decode!()
+        |> Kernel.get_in(["data", "images"])
+        |> List.first()
+        |> Kernel.get_in(["id"])
+
+      image_uri_for_imgur_id(image_id)
+    else
+      _err -> nil
+    end
+  end
+
+  defp image_uri_for_imgur_id(image_id) do
+    with headers when is_list(headers) <- get_imgur_headers() do
+      imgur_api_url = "https://imgur-apiv3.p.mashape.com/3/image/#{image_id}"
+      imgur_resp = retry_get_until_success(imgur_api_url, headers)
+
+      imgur_resp.body
+      |> Poison.decode!()
+      |> Kernel.get_in(["data", "link"])
+      |> URI.parse()
+    else
+      _err -> nil
+    end
+  end
+
+  defp retry_get_until_success(url, headers \\ []) do
+    retry with: exp_backoff() |> randomize |> cap(5_000) |> expiry(10_000) do
+      case HTTPoison.get!(url, headers) do
+        %{status_code: status_code} = resp when status_code in @success_status_codes ->
+          resp
+
+        resp ->
+          raise "Retry error for #{url} #{resp}"
+      end
+    after
+      result -> result
+    else
+      error -> error
+    end
+  end
+
+  # Performs a GET until we get response with a valid status code from ytdl.
+  #
+  # Oddly, valid status codes are 200, 302, and 500 since that indicates
+  # the server successfully attempted to resolve the video URL(s). If we get
+  # a different status code, this could indicate an outage or error in the
+  # request.
+  #
+  # https://youtube-dl-api-server.readthedocs.io/en/latest/api.html#api-methods
+  defp retry_get_until_valid_ytdl_response(url) do
+    retry with: exp_backoff() |> randomize |> cap(5_000) |> expiry(10_000) do
+      case HTTPoison.get!(url) do
+        %{status_code: status_code} = resp when status_code in @ytdl_valid_status_codes ->
+          resp
+
+        resp ->
+          raise "Retry error for #{url} #{resp}"
+      end
+    after
+      result -> result
+    else
+      error -> error
+    end
+  end
+
+  defp get_root_host(nil) do
+    nil
+  end
+
+  defp get_root_host(host) do
+    # Drop subdomains
+    host |> String.split(".") |> Enum.slice(-2..-1) |> Enum.join(".")
```

Comment:
Would this break for URLs like http://jenbartel.tumblr.com/image/175384943654 ?

Reply:
this is grabbing the last two parts of the domain name, so it will return `tumblr.com`

Reply:
Right, but tumblr.com/image/175384943654 is a 404

Reply:
ah, yeah the `root_host` that is extracted here is not actually used for the underlying requests -- it's just use as a way to pattern match the various handlers